### PR TITLE
Add `{into,from}_raw` to Rc and Arc

### PR DESCRIPTION
**Pull request by cristicbz**
These methods convert to and from a `*const T` for `Rc` and `Arc` similar to the way they work on `Box`. The only slight complication is that `from_raw` needs to offset the pointer back to find the beginning of the `RcBox`/`ArcInner`.

I felt this is a fairly small addition, filling in a gap (when compared to `Box`) so it wouldn't need an RFC. The motivation is primarily for FFI.

(I'll create an issue and update a PR with the issue number if reviewers agree with the change in principle **Edit: done #37197**) 

~~Edit: This was initially `{into,from}_raw` but concerns were raised about the possible footgun if mixed with the methods of the same name of `Box`.~~

Edit: This was went from `{into,from}_raw` to `{into,from}_inner_raw` then back to `{into,from}_raw` during review.

### CODE REVIEW

File: src/liballoc/arc.rs [hunk 1]
```diff
@@ -270,6 +270,81 @@ impl<T> Arc<T> {
...
+    /// ```
+    /// #![feature(rc_raw)]
+    ///
+    /// use std::sync::Arc;
+    ///
+    /// let x = Arc::new(10);
+    /// let x_ptr = Arc::into_raw(x);
+    /// assert_eq!(unsafe { *x_ptr }, 10);
+    /// ```
+    #[unstable(feature = "rc_raw", reason = "newly added", issue = "1")]
```
Comment (sfackler) [inline src/liballoc/arc.rs]: The `reason` field is no longer required, fyi.

**Comment by sfackler** [comment]
This seems reasonable to me I think - cc @rust-lang/libs.

**Comment by cristicbz** [comment]
Thanks @sfackler! I created the tracking issue, updated the PR with the issue number and removed the  reason field.

File: src/liballoc/arc.rs [hunk 2]
```diff
@@ -270,6 +270,81 @@ impl<T> Arc<T> {
...
+    ///
+    /// // The memory was freed when `x` went out of scope above, so `x_ptr` is now dangling!
+    /// ```
+    #[unstable(feature = "rc_raw", issue = "37197")]
+    pub unsafe fn from_raw(ptr: *const T) -> Self {
+        // To find the corresponding pointer to the `ArcInner` we need to subtract the offset of the
+        // `value` field from the pointer.
+
+        // Create a fake `ArcInner` to work with valid pointers only.
+        let invalid_arc_inner: ArcInner<T> = mem::zeroed();
```
Comment (alexcrichton) [inline src/liballoc/arc.rs]: Perhaps something like an [`offset_of!` macro](https://github.com/alexcrichton/ctest/blob/2839e49847a6adca6e96cc81c46a1f03f8562ac0/src/lib.rs#L731-L735) could be used instead of an arbitrary value here?

**Comment by alexcrichton** [comment]
Seems quite nifty to me! I like how it mirrors `Box` and I've wanted this myself before in a few odd situations.

Let's see if this works...

@rfcbot fcp merge

File: src/liballoc/arc.rs (hunk 2 above)
Comment (cristicbz) [inline src/liballoc/arc.rs]: I know that's a common pattern in C++, but isn't that UB? Just like `ptr.offset()` is UB if going more than one past end of allocation? That's the comment was meant to express. I could still wrap it in a  macro though, but I'm not sure where it'd go if I wanted to share it between `arc.rs` and `rc.rs`.

**Comment by diwic** [comment]
Thumbs up from me - I already to this for Arc in [RMBA](https://docs.rs/reffers/0.3.0/reffers/rmba/struct.RMBA.html) and if this was in stdlib I would be less fragile w r t changes to Arc's internal layout.

File: src/liballoc/arc.rs (hunk 2 above)
Comment (Amanieu) [inline src/liballoc/arc.rs]: [Here](https://github.com/Amanieu/intrusive-rs/blob/master/src/lib.rs#L232) is a safe version of `offset_of!` which doesn't rely on UB.

Comment (cristicbz) [inline src/liballoc/arc.rs]: @Amanieu thanks for the link! That's pretty much what I'm doing here (bar the clever check against `Deref`), do you think I should add your macro and use it for clarity and DRY though?

BTW: isn't this at risk of overflowing though: `val.$field as *const _ as isize`. I do the casts to `usize`, compute the difference and then cast to `isize`, since I thought the absolute address could well be greater than `isize::MAX`. Or is there some kind of guarantee that's not the case?

Comment (Amanieu) [inline src/liballoc/arc.rs]: @cristicbz nice catch, you're probably right about the overflow issue, I'll fix that.

The reason I used `isize` is that I then pass the resulting value to `.offset` to recover the address of the parent object. Using `.offset` is better than performing arithmetic on `usize` since it uses an LLVM GEP instruction, which helps with alias analysis. See the [`container_of!`](https://github.com/Amanieu/intrusive-rs/blob/master/src/lib.rs#L267) macro later in the same file.

Comment (cristicbz) [inline src/liballoc/arc.rs]: Ah yes, I could cast to `isize` and use `ptr.offset(-offset)`, good point! My only other worry with your code (and the reason I used `mem::zeroed`) is that, technically reads of uninitialized memory are UB as well. There's a lang lawyer question about whether `&undef.field` is a read of `undef`, but i wouldn't risk it!

**Comment by cristicbz** [comment]
Thanks for the feedback @Amanieu & @alexcrichton! I moved the `offset_of!` code into a macro using `ptr::offset` instead of my `usize` addition. Also borrowed comments and the `Deref`-preventing pattern trick from @Amanieu's library (hope that's OK!).

I did keep my casts to `usize` to avoid overflow and kept `mem::zeroed` over `mem::uninitialized` to avoid UB potentially caused by a very pedantic understanding of 'reading undef'.

Didn't really know where to put these macros so I dropped them unexported in a `liballoc/macros.rs` module. Happy to move it anywhere else, of course!

File: src/liballoc/rc.rs [hunk 3]
```diff
@@ -363,6 +363,68 @@ impl<T> Rc<T> {
...
+    ///     // Further calls to `Rc::from_raw(x_ptr)` would be memory unsafe.
+    /// }
+    ///
+    /// // The memory was freed when `x` went out of scope above, so `x_ptr` is now dangling!
+    /// ```
+    #[unstable(feature = "rc_raw", issue = "37197")]
+    pub unsafe fn from_raw(ptr: *const T) -> Self {
+        // To find the corresponding pointer to the `RcBox` we need to subtract the offset of the
+        // `value` field from the pointer.
+        Rc { ptr: Shared::new(ptr.offset(-offset_of_unsafe!(RcBox<T>, value)) as *mut _) }
```
Comment (cristicbz) [inline src/liballoc/rc.rs]: Argh yes another reason i was doing usize arithmetic before!

Comment (cristicbz) [inline src/liballoc/rc.rs]: (thanks for the heads up btw! :) )

Comment (cristicbz) [inline src/liballoc/rc.rs]: Done. Changed the name to `unoffset_field_ptr` since outside of intrusive containers `container_of` didn't seem that semantic for the operation.

**Comment by alexcrichton** [comment]
This seems... overly complicated to me, but it's just an internal detail so seems fine.

**Comment by cristicbz** [comment]
@alexchrichton Yeah it did get a bit out of hand... Could simplify down to an `offset_of` macro + a `(ptr as *const u8) .offset(-offset_of(RcBox<T>, data)) as *mut_` if you think that'd be a clearer.

**Comment by alexcrichton** [comment]
Perhaps yeah, I find `offset_of` hard to follow here. I feel like in practice it's not UB to use null pointers and such because this is how I've always seen that macro defined in C ...

**Comment by Amanieu** [comment]
@alexcrichton A C version of the macro that dereferences a NULL pointer is still UB in C. This is why gcc defines `offsetof` as an alias for `__builtin_offsetof`, which avoids the UB.

**Comment by cristicbz** [comment]
I pushed a simplified version which defines only the `offset_of` macro (as previously defined, using `mem::zeroed` and casts to `usize`) and inlines the `(ptr as *const u8).offset(-offset_of!(RcBox<T>, value)) as *mut _)` logic in `{A,}Rc::from_raw` implementations. I think this strikes a good balance, this was meant to be a very minimal change.

File: src/liballoc/macros.rs [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// Private macro to get the offset of a struct field in bytes from the address of the struct.
+//
+// This macro is identical to `offset_of!` but doesn't give a warning about unnecessary unsafe
+// blocks when invoked from unsafe code.
```
Comment (Amanieu) [inline src/liballoc/macros.rs]: You should probably fix this comment

Comment (cristicbz) [inline src/liballoc/macros.rs]: Whoops, didn't do a last `git add` to that file, thanks!

**Comment by aturon** [comment]
@fcp concern Footgun

I'm a little bit worried that this API is a footgun: the types and names strongly suggest that the only relevant raw representation is the `T` itself. 

At the very least, it's a violation of the usual convention for `raw`, which is to give you the *complete* raw representation of an object. Perhaps `into_inner_raw`/`from_inner_raw` would be enough to signal what's happening?

**Comment by aturon** [comment]
@rfcbot fcp concern Footgun

I'm a little bit worried that this API is a footgun: the types and names strongly suggest that the only relevant raw representation is the `T` itself. 

At the very least, it's a violation of the usual convention for `raw`, which is to give you the *complete* raw representation of an object. Perhaps `into_inner_raw`/`from_inner_raw` would be enough to signal what's happening?

**Comment by CryZe** [comment]
I agree with the change to `from_inner_raw`. I recently wanted to create an "Arced slice", so I created a vector, turned it into a boxed slice, called `Box::into_raw` on that and would've called `Arc::from_raw` on it. Fortunately it didn't exist yet, cause that would've caused some bugs, that could be hard to find. Unfortunately this PR still doesn't provide any way to create an "Arced slice", but the additional indirection isn't too bad.

**Comment by SimonSapin** [comment]
I agree with @aturon and @CryZe that types compatible with `Box::{into,from}_raw` but with different semantics (and memory allocator possibly relying on these semantics to avoid undefined behavior) is a footgun.

When I first thought of this feature, in my mind the `into_raw` methods returned `*const ArcInner<T>` and `*const RcBox<T>`. (Or perhaps `*mut` like `Box::into_raw` and `Box::from_raw`? Does it matter?)

`ArcInner` and `RcBox` would be made public, but with their fields kept private and no public API of their own. Probably one or both of them should be renamed so that their naming is consistent, but I don’t have an opinion on what name it should be.

**Comment by cristicbz** [comment]
@SimonSapin If I understand your suggestion correctly, exposing `ArcInner` and `RcBox` ptrs without public fields would defeat the purpose of this change for my use case: roundtripping `Rc` as `*T` pointers through FFI (in particular this arose in an `Rc<somelib_buffer_t>` scenario where API methods operate on `*const somelib_buffer_t` and they allow a callback to free the buffer).

@aturon I can see the argument for the `_inner_` variants and I'd be OK with it if that ends up being the consensus. Some counterarguments to consider:

  1. **Footgun because it's incompatible with `Box<T>`**. It's already the case that using any other pointer for `Box::from_raw` than one obtained using `Box::into_raw` is UB as the docs make clear, if nothing else because of allocator concerns. Something like `Vec::from_raw_parts(Box::new(10).into_raw(), 1)`. Obviously the extra length parameter makes it less of a footgun, but my point is more that "`Foo::from_raw` is only compatible with `Foo::into_raw`" seems to be an already accepted rule.

  2. **`Foo::into_raw` should return the entire representation of `Foo`**. I'd argue that it does! Bear with me:
   * An `Rc`'s representation **is** a single pointer: unlike the length in a slice's fat pointer, the ref count is not part of the `Rc` pointer, but is instead part of the allocation. I see the `RcBox` like `jemalloc`'s allocation book-keeping information.
   * You could imagine an alternative implementation for `Rc` where a thread local `HashMap<*const T, RefCounts>` is used. This is what the semantics correspond to. The fact that the pointer->count mapping is performed via a pointer offset (i.e. member access) can be seen as an implementation detail.
   * This view also suggests that the footgun issue translates to the common "`free`-ing a pointer via a different allocator than the one it was `malloc`-d with".

All this said, I _am_ sympathetic to the footgun argument. Just because we tell people not to pass an `Box::into_raw` result to an `Rc::from_raw` doesn't mean they won't (though the functions are unsafe so people should read the docs!).

I'd suggest `into_rc_raw` and `from_rc_raw` rather than `inner` to hammer home the fact that the methods are only compatible with `Rc`-s. Would that be acceptable?

**Comment by eternaleye** [comment]
@cristicbz 
> - An `Rc`'s representation is a single pointer: unlike the length in a slice's fat pointer, the ref count is not part of the `Rc` pointer, but is instead part of the allocation. I see the `RcBox` like `jemalloc`'s allocation book-keeping information.

An Rc's representation is a single raw pointer to an `RcInner<T>`, not to a `T`. An `RcInner<T>` may not meet your needs (FFI), but that's not the same as it not being there.

> - You could imagine an alternative implementation for `Rc` where a thread local `HashMap<*const T, RefCounts>` is used. This is what the semantics correspond to. The fact that the pointer->count mapping is performed via a pointer offset (i.e. member access) can be seen as an implementation detail.

You _could_ imagine that, yes. However, that _would_ have a different representation - a representation is part of the _concrete implementation_ of a type, not the _abstract behavior_ (and you're arguing that the latter being the same makes the former irrelevant).

**Comment by cristicbz** [comment]
@eternaleye First, apologies if my previous comment came across as combative. As I said, I'm very open to an alternative and I was just expanding on my reasoning for the approach I took in this PR. 

> An Rc's representation is a single raw pointer to an RcBox<T>, not to a T. An RcBox<T> may not meet your needs (FFI), and it may be non-public, but that's not the same as it not being there - and unlike jemalloc's bookkeeping information, it really is what the pointer is to.

That's true, I was trying to draw a distinction between something like a fat pointer (where the length or vtable is part of the pointer) versus something like `Rc` where the extra information is contained in the allocation---similar to a `malloc` implementation's book keeping information. Though there are differences, in my view they are more similar than different, but that's ultimately subjective.

> You could imagine that, yes. However, that would have a different representation - a representation is part of the concrete implementation of a type, not the abstract behavior (and you're arguing that the latter being the same makes the former irrelevant).

I think this may come down to the fact that I see a pointer to `T` and a pointer to a `BookKeeping<T>` as pretty much the same representation. Indeed, if `value` were the first field of `RcBox<T>`, then the representation would be bit-for-bit equivalent. Or if an `Rc` stored `*const T` internally, using it for the `Deref` implementations and only accessed the ref counts via a `transmute` to `RcBox<T>` for `clone()`. The constant in all these cases is that `Rc` wraps---one way or another---a pointer to `T`, with the invariant that some method of obtaining ref-counts from that pointer exists.


I don't want to spend too much energy on debating the semantics of 'representation'; this is ultimately a pragmatic matter for me: converting to and from `Rc`-s and raw pointers is useful. Our options seem to be:
  1. Rename the methods to something like `into_(rc|inner|rc_inner)_raw`.
  2. Expose `RcBox` with a public `value: T` field to support the `offset_of` shenanigans.
  3. Not support this use-case at all, because the trade-offs are not worth it.

When all's said and done, I have a workaround in place. If (3) is the consensus, feel free to close this PR.

**Comment by SimonSapin** [comment]
The use case in Stylo is that we `transmute()` from `Arc<T>` to a raw pointer that we give to C++ code through FFI. From the C++ point of view, that pointer is opaque and doing anything with it requires calling back into Rust which transmutes the pointer back to `Arc<T>`.

This PR would help avoid using `transmute` and making assumptions about the memory representation (that I believe is currently an internal implementation detail) of `Arc<T>`.

@cristicbz It sounds like you have a stronger use-case: you need the other side of the FFI to dereference that pointer to access a `T` value. Is that correct?

**Comment by cristicbz** [comment]
@SimonSapin Yeah, here's [the kind](http://api.zeromq.org/4-0:zmq-msg-init-data) of functions we're using:

```c
  int zmq_msg_init_data (zmq_msg_t *msg, void *data, size_t size, zmq_free_fn *ffn, void *hint);
```

> If provided, the deallocation function ffn shall be called once the data buffer is no longer required by ØMQ, with the data and hint arguments supplied to zmq_msg_init_data().

`*data` needs to be a valid `*const u8` which will be freed using `ffn`. `ffn` is passed simply the data pointer and is supposed to free it. In my case, I'm sharing these buffers using `Arc` so my `ffn` needs to convert back to an `Arc` from a pointer (I'm passing the length as hint and doing some trickery to avoid a fat pointer, but it's not really relevant).

File: src/liballoc/macros.rs [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    ($container:path, $field:ident) => {{
+        // Make sure the field actually exists. This line ensures that a compile-time error is
+        // generated if $field is accessed through a Deref impl.
+        let $container { $field : _, .. };
+
+        // Create an (invalid) instance of the container and calculate the offset to its
+        // field. Use `zeroed` rather than `uninitialized` to avoid an arguably pedantic read of
+        // uninitialized memory (`&undef.field`). Similarly using a null pointer might be UB
+        // if `&(*(0 as *const T)).field` if interpreted to be nullptr deref.
+        let invalid: $container = ::core::mem::zeroed();
```
Comment (arielb1) [inline src/liballoc/macros.rs]: Use `mem::uninitialized` rather than `mem::zeroed` - less likely to result in compiler violence.

Comment (cristicbz) [inline src/liballoc/macros.rs]: Done, thanks. I expected the opposite to be true. Compilers: how do they work?

**Comment by cristicbz** [comment]
@aturon @brson I changed the names to the `inner` variants, is that enough to unblock this? Or do you remain unconvinced on the philosophical issue of `Rc`-s representation?

**Comment by aturon** [comment]
@cristicbz Sorry for the slow reply -- and yes, I'm fine going forward with the renamed API.

**Comment by brson** [comment]
I'm hesitant about this.

Although I understand why it's named differently from `into_raw`, I don't think the distinction will be obvious to users - it just looks inconsistent. And I agree with @cristicbz's earlier point that it's just bogus to mismatch `into_raw` / `from_raw` calls with the wrong container types; I don't see making the names different helps clarify that contract.

But I also don't understand what the use case for wanting to pull out an unsafe pointer to the contained data while also destroying the `Rc` value. With `Box`s unique ownership it's obvious but `Rc` is a shared type so semantically there's no need to make the safe version go away and avoid touching it.

Furthermore, this method doesn't give unsafe code anyway to participate in the reference counting because it discards any pointer to it. Are we someday going to be called on to add `into_raw_parts` and do this over again?

**Comment by eternaleye** [comment]
To build on what @brson said - couldn't your use case be solved using the API @SimonSapin described like so?

```C
int zmq_msg_init_data (
    zmq_msg_t *msg,
    void *data, // Actually a *const ArcInner<T>
    size_t size,
    zmq_free_fn *ffn, // would call unref_arcinner(data), which would be an extern "C" fn written in Rust
    void *hint // Can still use hint to avoid fat pointers
);
```

To actually use the data, you'd call an `extern "C" fn` that basically just wrapped pulling the `*const T` out of the `*const ArcInner<T>`
When you want to

**Comment by cristicbz** [comment]
@eternaleye `void *data` needs to be a valid pointer to a buffer since it gets dereferenced by the C code, passing in a `*const ArcInner<T>` would read the reference counts from `ArcInner` before the actual data. Unless I'm missing something?

**Comment by eternaleye** [comment]
@cristicbz: Oh, right, I see now - it's been a while since I used the ZMQ API; sorry.

Having said that, is it actually valid to call `zmq_msg_init_data` on an `Arc`'s data at all? The API says it "takes ownership of the supplied buffer", which seems like it's entirely incompatible with `Arc`.

In particular, wouldn't it be a const-correctness violation, as the ZMQ api makes no guarantees about writing to the buffer?

**Comment by cristicbz** [comment]
@eternaleye You're right that's a potential risk and you need to be careful about what 0MQ functions you call on that message! If you're just sending the message 0MQ doesn't overwrite the buffer (it just reads it). If you want to share a constant frame (like a tag or a ROUTER's routing frame) between multiple messages this is a useful technique.

**Comment by alexcrichton** [comment]
@brson 

> But I also don't understand what the use case for wanting to pull out an unsafe pointer to the contained data while also destroying the Rc value. With Boxs unique ownership it's obvious but Rc is a shared type so semantically there's no need to make the safe version go away and avoid touching it.

This use case can be covered by a C API expecting `*const T`and you have an `Rc<T>`, right? That is, you could have a `Rc` pointer which you're passing ownership of one reference to into a C API. Later you're called to free that data, where you call `from_raw`, and it may or may not deallocate memory.

One concrete use case I've had for this in the past is in mio itself on the Windows bindings. There there's a struct that looks like `Arc<MyOverlapped>` where `MyOverlapped` starts with an `OVERLAPPED` structure (from winapi). Ownership of one reference is then passed to IOCP, but that's expecting a `*mut OVERLAPPED` (mutability can be ignored here). Later on once I get back the `*mut OVERLAPPED` I know how to drop the reference.

mio currently vendors its own arc implementation, but this would actually be perfect for replacing that!

> Furthermore, this method doesn't give unsafe code anyway to participate in the reference counting because it discards any pointer to the reference count itself. Are we someday going to be called on to add into_raw_parts or some similar method and do this over again?

I don't think we'd need to add `into_raw_parts`, we could have something like this right?

```rust
let my_rc: *const T = ...;
let my_rc_reified = Rc::from_raw(my_rc);
let my_clone = my_rc_reified.clone();
let my_rc_again = Rc::into_raw(my_rc_reified);
```

That is, if you want to modify the reference count, you'd cast it back to a `Rc<T>` and then back again once the operation is done.

**Comment by brson** [comment]
> This use case can be covered by a C API expecting *const Tand you have an Rc<T>, right? That is, you could have a Rc pointer which you're passing ownership of one reference to into a C API. Later you're called to free that data, where you call from_raw, and it may or may not deallocate memory.

Yes. I was thinking you might as well just stash an Rc in safe code and hold ownership on behalf of the C code, but I understand wanting to give a share of the ownership to C.

> That is, if you want to modify the reference count, you'd cast it back to a Rc<T> and then back again once the operation is done.

If you have access to Rust. You can't use that to let C manage reference counts, which is more what I was thinking.

**Comment by rkruppe** [comment]
> If you have access to Rust. You can't use that to let C manage reference counts, which is more what I was thinking.

That would require a `#[no_mangle] extern "C" fn`. That's pretty unprecedented in `std`. There's also the problem that it would either have to be generic (non-starter for being callable from C) or pick some fixed but wrong type for the `Rc` contents &mdash; I don't dare claim that this will be UB, but it will at least be extremely gross. Luckily, none of this is necessary, since any Rust crate can do the `from_raw`-`clone`-`to_raw` dance in its own `#[no_mangle] extern "C"` helper function if and when the C code it interacts with needs it, for whatever concrete types it deals with.

**Comment by brson** [comment]
Is `*const` really the correct type here? Arc's and Rc's are often filled with mutable things. And `Shared`, the type used to implement `Rc`, is for managing `*mut` (though weirdly actually contains a `*const`).

**Comment by brson** [comment]
> That would require a #[no_mangle] extern "C" fn

Or a pointer to the refcount.

My objections here are pretty squishy, that's its slightly different than other things, in naming and in the magicy implementation. But I understand the use case is real and we should offer some solution. I'd like an answer on `*const` vs `*mut` before accepting the fcp though.

**Comment by rkruppe** [comment]
> Or a pointer to the refcount.

That may work for increasing the ref count (though I would rather not rely on random C code everywhere to correctly reproduce the atomic instruction necessary), but what about decreasing it? How do you invoke the destructor? (I just noticed that's also a problem for the "`std` provides a single `#[no_mangle] extern "C" fn`" approach but that solution was a straw man anyway.)

**Comment by alexcrichton** [comment]
> Yes. I was thinking you might as well just stash an Rc in safe code and hold ownership on behalf of the C code, but I understand wanting to give a share of the ownership to C.

At least with the use case I had in mio, it was perf sensitive, so stashing the pointer in safe code would have required extra storage requirements where passing ownership to C didn't require any extra data structures. In that sense I do think this'd be possible for many cases, but perhaps not all.

> If you have access to Rust. You can't use that to let C manage reference counts, which is more what I was thinking.

Ah yeah that's true, but I'm not sure that C would ever expect to be able to do this? At least, I can't really think of a use case for that.

> Is *const really the correct type here?

Hm yeah that's a good point, I might personally expect `*mut`

---

FWIW I slightly prefer the name `into_raw` as opposed to `into_inner_raw` like @brson, but I don't feel too strongly either way

**Comment by cristicbz** [comment]
> Hm yeah that's a good point, I might personally expect *mut

I changed this to `*mut` now, I'm not too fussed since mutability on raw ptr-s is a guideline at best :P.

My reasoning for `*const T` was that, mutating the value through the returned pointer breaks aliasing rules for any remaining `Rc`-s which expect the data they point to to be immutable. If a C function expects a non-const pointer that should be a warning sign that passing in the result of `Rc::into_raw` might not be correct.

> FWIW I slightly prefer the name into_raw as opposed to into_inner_raw like @brson, but I don't feel too strongly either way

So do I, but I don't feel to strongly either---it was @aturon who requested it.

**Comment by aturon** [comment]
I'm willing to be overruled on the name. Let's just get this thing landed. :)

**Comment by alexcrichton** [comment]
@brson, do you feel comfortable checking your box with an updated name?

**Comment by cristicbz** [comment]
I changed this back to `{into,from}_raw`, let me know if there's anything else!

**Comment by cristicbz** [comment]
@aturon @alexchrichton @brson ping?

**Comment by alexcrichton** [comment]
@cristicbz just waiting on @brson to check [his box](https://github.com/rust-lang/rust/pull/37192#issuecomment-254030846)

**Comment by cristicbz** [comment]
Amazing, thanks for putting up with me and getting this merged!

**Comment by cristicbz** [comment]
(also given the relative "controversy", I'd suggest relnotes may be appropriate to get another batch of input when TWiR comes around)

**Comment by aturon** [comment]
@cristicbz 

> Amazing, thanks for putting up with me and getting this merged!

No, thank **you** for sticking with it despite the lengthy process! Hope to see lots more to come.

**Comment by rkjnsn** [comment]
I find @cristicbz's reasoning for initially going with `*const T` convincing, and am not sure how we ended up with `*mut T`. @brson says "Arc's and Rc's are often filled with mutable things", but as far I as can tell, this is only true when `T` has interior mutability of some sort. (The exception is `get_mut`, which is only usable if no other references exist.)

File: src/liballoc/macros.rs [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// Private macro to get the offset of a struct field in bytes from the address of the struct.
+macro_rules! offset_of {
+    ($container:path, $field:ident) => {{
+        // Make sure the field actually exists. This line ensures that a compile-time error is
+        // generated if $field is accessed through a Deref impl.
+        let $container { $field : _, .. };
```
Comment (eddyb) [inline src/liballoc/macros.rs]: This could've used `ref $field` to avoid using field access below (came here from the tracking issue which I found from the docs).
